engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/blueprint/NadelExecutionBlueprintFactory.kt

```diff
@@ -567,3 +599,138 @@ private class SharedTypesAnalysis(
         }
     }
 }
+
+private class TypeOwnership(
+    private val overallSchema: GraphQLSchema,
+    private val services: List<Service>,
+    private val typeRenames: List<NadelTypeRenameInstruction>,
+) {
+    fun getTypesToServicesThatDefineThem(): Map<String, Set<Service>> {
+        return getTypesThatServiceDefines()
+            .entries
+            .asSequence()
+            .flatMap { (service, typesDefined) ->
+                typesDefined
+                    .asSequence()
+                    .map {
+                        it to service
+                    }
+            }
+            .groupBy {
+                it.first
+            }
+            .mapValues { it: Map.Entry<String, List<Pair<String, Service>>> ->
+                it.value // Get the List<Pair<…>> out
+                    .asSequence()
+                    .map {
+                        it.second // Get the Service out of the Pair<String, Service>
+                    }
+                    .toSet()
+            }
+    }
+
+    fun getTypesThatServiceDefines(): Map<Service, Set<String>> {
+        return services
+            .map { service ->
+                service to getTypeThatServiceDefined(service)
+            }
+            .toMapStrictly()
+    }
+
+    private fun getTypeThatServiceDefined(service: Service): Set<String> {
+        val typesDefined = service.definitionRegistry.definitions
+            .asSequence()
+            .filterIsInstance<AnyNamedNode>()
+            .filterNot { it.isExtensionDef }
+            .map { it.name }
+            .toHashSet()
+
+        val rawTypesUsedAsOutputTypes = service.definitionRegistry.definitions
```

Comment:
Looks at all the types the service uses as output types for fields.

e.g. consider `PageInfo` which is defined in the shared service. Every service uses it, but it will never show up it their `Service.definitionRegistry` because it has already been defined once in the overall schema. We need to include that in the types the service defines.